### PR TITLE
Add decaying average logic to the frontend

### PR DESCRIPTION
**Pull request by NealGeilen**
#83 needs to be merged first.

### CODE REVIEW

**Comment by NealGeilen** [comment]
When selecting a term, the decaying average does not change. 
Is this something that we want? Or do we want that the decaying average to change depending on the selected term?
This will change the results and thus create confusion when things are "masterd" 

@Typiqally @jendafojtik

**Comment by Typiqally** [comment]
> When selecting a term, the decaying average does not change. Is this something that we want? Or do we want that the decaying average to change depending on the selected term? This will change the results and thus create confusion when things are "masterd"
> 
> @Typiqally @jendafojtik

The decaying average should change when selecting a term. It is important that it reflects what the decaying average would have looked like at that time in the term, so that a teacher (or student) can essentially go back in time.

**Comment by honzafojtik** [comment]
@Typiqally @NealGeilen I'll look over the formula this weekend and make sure to have it worked out by Monday!

**Comment by NealGeilen** [comment]
Recording 2023-05-09 000602

The current state of development.

Up next is implementing the personal development graph

**Comment by NealGeilen** [comment]
@Typiqally Ready for review

File: Epsilon.Abstractions/Epsilon.Abstractions.csproj [hunk 1]
```diff
@@ -10,4 +10,8 @@
       <ProjectReference Include="..\Epsilon.Canvas.Abstractions\Epsilon.Canvas.Abstractions.csproj" />
     </ItemGroup>
 
+    <ItemGroup>
+      <Folder Include="Component\" />
+    </ItemGroup>
+
```
Comment (Typiqally) [inline Epsilon.Abstractions/Epsilon.Abstractions.csproj]: Usually this gets added if this directory is empty, if it is empty, please remove it

File: Epsilon.Abstractions/Model/OutcomeResult.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Epsilon.Abstractions.Model;
+
+public abstract record OutcomeResult
```
Comment (Typiqally) [inline Epsilon.Abstractions/Model/OutcomeResult.cs]: Can you put the required properties in the constructor

File: Epsilon.Host.Frontend/src/components/PersonalDevelopmentGraph.vue [hunk 3]
```diff
@@ -76,22 +81,43 @@ const chartOptions = {
...
+    if (props.domain.masteryLevels == null) {
+        return undefined
+    }
+
+    return props.domain.masteryLevels.find(
+        (masteryLevel) => masteryLevel.id == masteryId
+    )
+}
+
+watch(() => loadChartData())
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/components/PersonalDevelopmentGraph.vue]: Can you move this under `onMounted`?

File: Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    /**
+     * Calculate decaying average described by Canvas: https://community.canvaslms.com/t5/Canvas-Basics-Guide/What-are-Outcomes/ta-p/75#decaying_average
+     * !IMPORTANT, The list of results will always have to be a list of the same outcome id. Not a list of equal activity and architecture layer.
+     * @param results
+     * @constructor
+     * @private
+     */
+    private static GetDecayingAverageFromOneOutcomeType(
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts]: Naming convention, please use camel case

File: Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    /**
+     * Calculate average of given tasks divided in architectural layers
+     * @param taskResults
+     * @param domain
+     * @constructor
+     * @private
+     */
+    private static GetDecayingAverageForAllOutcomes(
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts]: Camel case

File: Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        }) as DecayingAveragePerSkill[]
+    }
+
+    /**
+     * Calculate the averages for each task type divided in architecture layers.
+     * @param taskResults
+     * @param domain
+     * @constructor
+     */
+    public static GetAverageTaskOutcomeScores(
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts]: Camel case

File: Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+export class DecayingAverageLogic {
+    /**
+     * Calculate the averages for each skill type
+     * @param taskResults
+     * @param domain
+     * @constructor
+     */
+    public static GetAverageSkillOutcomeScores(
```
Comment (Typiqally) [inline Epsilon.Host.Frontend/src/logic/DecayingAverageLogic.ts]: Camel case

**Review by Typiqally — CHANGES_REQUESTED** [review]

**Review by Typiqally — CHANGES_REQUESTED** [review]
Well done, some minor things

File: Epsilon.Host.Frontend/src/components/PersonalDevelopmentGraph.vue (hunk 3 above)
Comment (NealGeilen) [reply Epsilon.Host.Frontend/src/components/PersonalDevelopmentGraph.vue]: Apparently the watch is not triggered when placed inside the mount function

**Review by Typiqally — APPROVED** [review]

**Pull request merged**